### PR TITLE
Add Dockerfile, Makefile rule, Goreleaser, and Github action

### PR DESCRIPTION
**Pull request by davidallendj**
This PR adds a Dockerfile to containerize the configurator. The container includes the main driver binary executable as well as the default plugins built with the project. It also includes a new Makefile rule that should make it easy to build by simply running `make docker`.

### CODE REVIEW

**Comment by alexlovelltroy** [comment]
Can you add a GitHub action and goreleaser config for this service as well?

**Review by synackd — CHANGES_REQUESTED** [review]
I made some Makefile comments, some of which need to be addressed (hence requesting changes), some of which are suggestions for more flexibility.

All of the fixes and recommendations can be summarized by a modified Makefile that I got working as follows:

```make
# Unless set otherwise, the container runtime is Docker
DOCKER ?= docker

prog ?= configurator
sources := main.go $(wildcard cmd/*.go)
plugin_source_prefix := pkg/generator/plugins
plugin_sources := $(wildcard $(plugin_source_prefix)/*/*.go)
plugin_binaries := $(addprefix lib/,$(patsubst %.go,%.so,$(notdir $(plugin_sources))))

# build everything at once
.PHONY: all
all: plugins exe test

# build the main executable to make configs
.PHONY: main driver binaries exe
main: exe
driver: exe
binaries: exe
exe: $(prog)

# build named executable from go sources
$(prog): $(sources)
        go build --tags=all -o $(prog)

container: binaries plugins
        $(DOCKER) build . --build-arg --no-cache --pull --tag 'configurator:test'

container-testing: binaries plugins
        $(DOCKER) build . --tag configurator:testing

# build all of the generators into plugins
.PHONY: plugins
plugins: $(plugin_binaries)

# how to make each plugin
lib/%.so: pkg/generator/plugins/%/*.go
        mkdir -p lib
        go build -buildmode=plugin -o $@ $<

# remove executable and all built plugins
clean:
        rm -f configurator
        rm -f lib/*

# run all of the unit tests
test:
        go test ./tests/generate_test.go --tags=all
```

**NOTE:** The Docker label and tag is hard-coded in the above and needs to be changed to be dynamic. I also didn't include some `.PHONY` directives that probably should be included.

The above provides some benefits:

- Modifying the Makefile is not required for adding/removing plugins
- Using a different container runtime other than docker is supported
- Things are only rebuilt when their sources change

**Comment by davidallendj** [comment]
> I made some Makefile comments, some of which need to be addressed (hence requesting changes), some of which are suggestions for more flexibility.
> 
> All of the fixes and recommendations can be summarized by a modified Makefile that I got working as follows:
> 
> ```makefile
> # Unless set otherwise, the container runtime is Docker
> DOCKER ?= docker
> 
> prog ?= configurator
> sources := main.go $(wildcard cmd/*.go)
> plugin_source_prefix := pkg/generator/plugins
> plugin_sources := $(wildcard $(plugin_source_prefix)/*/*.go)
> plugin_binaries := $(addprefix lib/,$(patsubst %.go,%.so,$(notdir $(plugin_sources))))
> 
> # build everything at once
> .PHONY: all
> all: plugins exe test
> 
> # build the main executable to make configs
> .PHONY: main driver binaries exe
> main: exe
> driver: exe
> binaries: exe
> exe: $(prog)
> 
> # build named executable from go sources
> $(prog): $(sources)
>         go build --tags=all -o $(prog)
> 
> container: binaries plugins
>         $(DOCKER) build . --build-arg --no-cache --pull --tag 'configurator:test'
> 
> container-testing: binaries plugins
>         $(DOCKER) build . --tag configurator:testing
> 
> # build all of the generators into plugins
> .PHONY: plugins
> plugins: $(plugin_binaries)
> 
> # how to make each plugin
> lib/%.so: pkg/generator/plugins/%/*.go
>         mkdir -p lib
>         go build -buildmode=plugin -o $@ $<
> 
> # remove executable and all built plugins
> clean:
>         rm -f configurator
>         rm -f lib/*
> 
> # run all of the unit tests
> test:
>         go test ./tests/generate_test.go --tags=all
> ```
> 
> **NOTE:** The Docker label and tag is hard-coded in the above and needs to be changed to be dynamic. I also didn't include some `.PHONY` directives that probably should be included.
> 
> The above provides some benefits:
> 
> * Modifying the Makefile is not required for adding/removing plugins
> * Using a different container runtime other than docker is supported
> * Things are only rebuilt when their sources change

These all look like good changes to make. I'm going to go ahead and add them and push.

File: Makefile
Comment (synackd) [inline Makefile:27]: For this target and the `container-testing` target, we can replace `configurator` with `$(prog)`. Do we want to hard-code `testing` for the version? Or should we put do something like `<git_tag>-dirty`?

Comment (davidallendj) [reply Makefile:27]: I think we should do `<git_tag>-dirty` for `container` and then something like `$(prog):<git_tag>-dirty`. I think that would probably make more sense.

File: Makefile
Comment (synackd) [inline Makefile:24]: I think you have spaces here instead of a tab. Getting:
```
Makefile:24: *** missing separator.  Stop.
```

Comment (davidallendj) [reply Makefile:24]: Copy-pasta fail, I guess.

File: Makefile
Comment (synackd) [inline Makefile:5]: You'll need to use the `shell` directive here. E.g:

```make
git_tag := $(shell git describe --abbrev=0 --tags)
```

File: Makefile
Comment (synackd) [inline Makefile:51]: I think it would be a good idea to have the plugins and main program binary by prerequisites of this target. That way, running `make test` on a fresh checkout would build first before testing. E.g:

```make
test: $(prog) $(plugin_binaries)
```

Comment (davidallendj) [reply Makefile]: I might change this so the main program is also being build in the testing directory. That way, we should be able to ensure that the main driver is built at the same time as the plugins.

**Comment by davidallendj** [comment]
Looks like that final change was made. Can I get another review on this @synackd?

File: Makefile
Comment (synackd) [reply Makefile:45]: Resolved in 73ca17.

**Review by synackd — APPROVED** [review]
Things look good, I just suggested a change that would prevent `git tag` from failing.

The `test` target is currently failing I think because it's using an old tag that has a function that doesn't exist. Merging and tagging this I think will fix that.

**Comment by synackd** [comment]
I don't think I have write access to this repo, otherwise it would allow merge.

**Comment by davidallendj** [comment]
> I don't think I have write access to this repo, otherwise it would allow merge.

You should have access now.

**Pull request merged**
